Solution/ronald ris with tests

ronaldris21 opened the pull request:
For this solution I've adapted the exercises in order to have test for the solutions

ronaldris21 committed:
Callback exercise solution

ronaldris21 committed:
1-Callback Solved - validate-user.js refractored to improve code reuse

ronaldris21 committed:
feat: problem 2 and 3 solved

ronaldris21 committed:
Readme Update: commands for running code

ronaldris21 committed:
Readme Update: commands for running code

ronaldris21 committed:
feat: unit test to solved problems

ronaldris21 committed:
fix: no users data on exercise 1, make sure it calls the callback and passed the tests

juanmdpc commented on 3-async-await/asyncAwait.test.js:
```diff
@@ -0,0 +1,28 @@
+
+describe("3. AsyncAwait:", () => {
+  it("74", async () => {
+    expect(await solution(74)).toEqual({
```
We can use 'toStrictEqual` to avoid allowing undefined properties of an object or array. Another alternative is use `toMatchObject` to only compare a set of properties of an object.

ronaldris21 replied:
It seems to be a better approach to use "toStrictEqual" on this tests. I'll change it. 
Thx

juanmdpc commented on 3-async-await/solution.js:
```diff
@@ -20,16 +20,129 @@ Example:
+  // YOUR SOLUTION GOES HERE
+  const promiseAllSolution = async (id) => {
+    try {
+      let resultsAll = await Promise.all([pricesAsync(id), productsAsync(id)]);
+      return {
+        id,
+        price: resultsAll[0],
+        product: resultsAll[1],
+      };
```
we can use destructuring here
```suggestion
      let [price, product] = await Promise.all([pricesAsync(id), productsAsync(id)]);
      return {
        id,
        price,
        product,
      };
```

ronaldris21 replied:
Awesome! promise.all returns directly the result values, so I really like this approach
I'll keep it mind

juanmdpc commented on 3-async-await/solution.js:
```diff
@@ -20,16 +20,129 @@ Example:
+    if (errorReasons) {
+      return errorReasons;
+    }
+
+    return {
+      id,
+      price: resultsAllSettled[0].value,
+      product: resultsAllSettled[1].value,
+    };
```
In order to be more clear in naming variables, we can use destructuring here
```suggestion
    const [priceSettled, productSettled] = resultsAllSettled
    
    return {
      id,
      price: priceSettled.value,
      product: productSettled.value,
    };
```

ronaldris21 replied:
Good for readability!
I've already applied this changes, and solved the gap it left after change the variable names.

I realized not all the solution methods have test, so I'll refractored the code too.

ronaldris21 committed:
test(asyncAwait): add individual tests for each promise method (all, allSettled, race, any)

Added separate tests for each Promise method (all, allSettled, race, any) in the asyncAwait exercise.

Fix(asyncAwait): applied PR recomendations

ronaldris21 commented:
Tests take a long time to run (around 30 seconds).
I'll try to run each test asynchronous to improve testing performance

ronaldris21 committed:
refractor(asyncAwait-tests): split the tests into separate files for each promise method (all, allsettled, any,race)

Performance improve at least 240%, being tests executed faster now

ronaldris21 committed:
Update readme: test commands samples

ronaldris21 committed:
refractor(asyncAwait): applying destructuring

juanmdpc commented:
Awesome job! Keep it up! 👊
